systemd/watch.go

```diff
@@ -0,0 +1,114 @@
+package systemd
+
+import (
+	"context"
+	"sync"
+
+	"github.com/virtual-kubelet/node-cli/provider"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+)
+
+// If a provider implements the Updater interface we will send pings when configmaps and secrets are updated.
+type Updater interface {
+	UpdateConfigMap(context.Context, *corev1.Pod, *corev1.ConfigMap) error
+	UpdateSecret(context.Context, *corev1.Pod, *corev1.Secret) error
+}
+
+type Watcher struct {
+	mu        sync.RWMutex
+	clientset *kubernetes.Clientset
+	// maps are indexed by "namespace.name"
+	configs map[string][]*corev1.Pod
+	secrets map[string][]*corev1.Pod
+}
+
+func newWatcher(clientset *kubernetes.Clientset) *Watcher {
+	return &Watcher{
+		configs:   make(map[string][]*corev1.Pod),
+		secrets:   make(map[string][]*corev1.Pod),
+		clientset: clientset,
+	}
+}
+
+func (w *Watcher) Run(p provider.Provider) error {
+	cmwatcher, err := w.clientset.CoreV1().ConfigMaps(corev1.NamespaceAll).Watch(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	swatcher, err := w.clientset.CoreV1().Secrets(corev1.NamespaceAll).Watch(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	for {
+		select {
+		case event, ok := <-cmwatcher.ResultChan():
+			if !ok {
+				cmwatcher, _ = w.clientset.CoreV1().ConfigMaps(corev1.NamespaceAll).Watch(context.TODO(), metav1.ListOptions{})
+				break
+			}
+			cm, ok := event.Object.(*corev1.ConfigMap)
+			if !ok {
+				continue
+			}
+			// event.Type == watch.Deleted ??
+			if event.Type != watch.Added && event.Type != watch.Modified {
+				continue
+			}
+
+			namespace := cm.ObjectMeta.Namespace
+			name := cm.ObjectMeta.Name
+			println("configmap update seen", namespace, name)
+			w.mu.RLock()
+			pods := w.configs[namespace+"."+name]
+			w.mu.RUnlock()
+			if u, ok := p.(Updater); ok {
+				for _, pod := range pods {
+					u.UpdateConfigMap(context.TODO(), pod, cm)
```

Comment:
Where is a filter so that this is called only for Pods that reference the updated ConfigMap?